23404 improve client role listing

sschu opened the pull request:
Fixes #23404 

First draft.

sschu commented:
First iteration on improving the situation here. I would rather use an iterative approach here rather than having a big bang rewrite also changing the admin API. The PR also tries to improve listing effective roles for a user as this also times out with many clients and it is essentially the same code. In the first iteration:

1. I improved the `RoleMapper.convertToModel()` method to  be O(1) instead of O(#clients) as it essentially made all listing operations O(#clients^2).
2. I improved the listing the effective role mapping to not be O(#clients) but O(#role-mappings).
3. I refactored the code to make it more readable and explicit, also because code sharing between available and effective role mappings won't be possible in the future (the former definitely has to go to the database). But this can be debated.

What is missing is an actual improvement to listing the available role mappings, this is as bad as before. This definitely has to look for roles that can be assigned so it has to somehow iterate over all roles. To make this efficient, this (and pagination) has to happen on the database level. This would be my next step. Other open questions:

1. The fact that only client roles are listed is due to the fact that the same for realm roles is taken from the admin api. For effective role mappings, this could easily include realm roles as well meaning the UI would only have to make one call (currently assigned realm roles are filtered out). This would also be possible for available roles.
2. `listCompositeRealmRoleMappings` doesn't do what it says it just lists subroles of the default role instead of the role provided as parameter. I don't know if this is intended but it looks super odd to me.
3. In the first iteration, I would probably leave everything in `admin-ui-ext`with the option of using it to improve the admin API later.

sschu commented:
With fine-grained permissions, the most efficient way would be to:
1. collect all roles the user can manage via fine-grained permissions directly
2. collect all roles the user can manage indirectly via fine-grained permissions on a client
3. subtract all roles has directly assigned
Ideally, this would all happen in the database so the database can directly do the pagination. Due to the nature of the user filtering by policies, I fear this won't be possible. So I would probably go for a similar approach to how fine-grained permissions are used when searching for users: add a method `getRolesWithMapPermissions()` to `RolePermissions.java` and use its output as input for a DB query with pagination.

thomasdarimont commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/RoleMappingResource.java:
```diff
@@ -21,23 +22,18 @@ public RoleMappingResource(RealmModel realm, AdminPermissionEvaluator auth) {
-                .skip(first).limit(max).collect(Collectors.toList());
+    static Stream<GroupModel> addParents(GroupModel group) {
+        //no cycle check here, I hope that's fine
+        if (group.getParent() == null) return Stream.of(group);
```
```suggestion
        if (group.getParent() == null) {
            return Stream.of(group);
        }
```

thomasdarimont commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/model/RoleMapper.java:
```diff
@@ -2,14 +2,14 @@
-        ClientModel clientModel = clients.filter(c -> roleModel.getContainerId().equals(c.getId())).findFirst()
-                .orElseThrow(() -> new IllegalArgumentException("Could not find referenced client"));
+        ClientModel clientModel = realm.getClientById(roleModel.getContainerId());
+        if (clientModel==null) throw new IllegalArgumentException("Could not find referenced client");
```
```suggestion
        if (clientModel==null) {
            throw new IllegalArgumentException("Could not find referenced client");
        }
```

thomasdarimont commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/AvailableRoleMappingResource.java:
```diff
@@ -3,6 +3,8 @@
 import java.util.Objects;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
```
```suggestion
```

thomasdarimont reviewed (changes requested):
Looks good to me, just a few minor remarks.

vramik reviewed (commented):
Thank you @sschu for this PR! It looks good to me, just few minor comments.

sschu commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/model/RoleMapper.java:
```diff
@@ -2,14 +2,14 @@
-
-    public static ClientRole convertToModel(RoleModel roleModel, Stream<ClientModel> clients) {
+    public static ClientRole convertToModel(RoleModel roleModel, RealmModel realm) {
         ClientRole clientRole = new ClientRole(roleModel.getId(), roleModel.getName(), roleModel.getDescription());
```
Thanks, will fix this.

sschu commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/RoleMappingResource.java:
I see the reason is that these variables are `private final` and overshadowed in the subclasses. Is there any specific reason for that? Otherwise I would just make them protected in the base class so they can be used in the subclasses.

vramik replied:
yes, that's the second option. Both options works for me, I leave it up to to decide what you prefer.

sschu commented:
Thanks for the feedback @vramik !
I am currently wondering if `getRolesStream` is actually the best name for the newly introduced method. In other places, methods taking a search string parameter are typically called `searchXyz`. It might even make sense to move the methods to the `RoleLookupProvider` interface (or even introduce a `RoleQueryProvider`? If I look at how it's done for Users, "lookup" typically seems to refer to methods for retrieving individual elements based on various parameters, not for searching for a set of elements).
Also, I am thinking of switching to the Criteria API to not bloat the `RoleEntity.java` with named functions and have some reuse. The search semantics is also not complete as searching for client roles should also include the client IDs, not just the role names.

sschu commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/EffectiveRoleMappingResource.java:
```diff
@@ -170,7 +192,11 @@ public final List<ClientRole> listCompositeUsersRoleMappings(@PathParam("id") St
         auth.roles().requireList(realm);
         final RoleModel defaultRole = this.realm.getDefaultRole();
-        return mapping(o -> o.hasRole(defaultRole)).collect(Collectors.toList());
+        //this definitely does not return what the descriptions says
```
@vramik What do you think about this endpoint here? It definitely does not do what the description says (also before my change).

vramik replied:
I agree the description looks a bit suspicious. Thanks for spotting this. I was trying to find out whether the description or the implementation is off, but I wasn't able to locate correct place from where the endpoint is called from UI.

@edewit could you please comment this one?

edewit replied:
Sorry, for missing this ping. Yes it is used in the UI

sschu replied:
@edewit But what is it used for? The descriptions says, it returns all client subroles for a realm role with a given id, the implementation actually returns all client subroles of the default realm role and the given id is not used at all.

edewit replied:
Right the unused ID is because that way we can make the UI always send one also when no ID is needed it's just to make this a more general implementation.

sschu replied:
Alright, thanks!

vramik commented:
Thank you @sschu!

> I am currently wondering if `getRolesStream` is actually the best name for the newly introduced method. In other places, methods taking a search string parameter are typically called `searchXyz`. It might even make sense to move the methods to the `RoleLookupProvider` interface (or even introduce a `RoleQueryProvider`? If I look at how it's done for Users, "lookup" typically seems to refer to methods for retrieving individual elements based on various parameters, not for searching for a set of elements). 

Good point, I think it makes sense to rename the method to `searchXzy` and move it to `RoleLookupProvider`. I wouldn't create new interface `RoleQueryProvider` for this.

> Also, I am thinking of switching to the Criteria API to not bloat the `RoleEntity.java` with named functions and have some reuse. 

Currently the jpa storage layer is not consistent in using queries. Whether it's named queries, criteria API or even native queries at some places. If you find it convenient, e.g. to have some reuse there, then go for it. I wouldn't mind the current approach with named queries as well.

> The search semantics is also not complete as searching for client roles should also include the client IDs, not just the role names. 

Would be the searching by client IDs required by / useful for by admin UI? Or do you mention it in sake of completeness?

sschu commented:
It's just the existing behaviour to search in client id and role name when searching for client roles, see https://github.com/keycloak/keycloak/blob/6736f319524ad268bd265cace7a27d67466453bb/rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/RoleMappingResource.java#L35.

sschu commented:
I updated the implementation so listing effective role mappings sorts correctly and available role mappings are listed using the new function from the `RoleLookupProvider` for the case that fine-grained permissions are disabled.  Next step is to implement listing available role mappings with fine-grained permissions enabled.
Some open questions for me:
- I added the new search method as a default method so alternative `RoleLookupProviders` still compile. However, they will encounter a runtime exception if they don't implement the method so this might not be ideal.
- Searching for available client roles in the `RoleLookupProvider` searches for matches and role name and client name because that's what the current API is doing to list available client roles. However, it is not consistent with what the other search functions are doing, they search in role name and role description.
- Listing effective role mappings doesn't offer pagination (it was like that before) but the admin console actually sends a request with pagination, at least when listing roles for a user.
- Listing effective role mappings could probably be more efficient by going to the DB directly similar to available role mappings. But it might not be that important considering that the amount of assigned roles should not be that big.

WDYT? @vramik  @danielFesenmeyer  @thomasdarimont

sschu committed:
23404 more efficient listing of assigned and available client role mappings

Signed-off-by: Sebastian Schuster <sebastian.schuster@bosch.io>
Co-authored-by: Vlasta Ramik <vramik@users.noreply.github.com>

sschu commented:
I finished the implementation to also work with fine-grained permissions. There are still some open issues but I would rather handle them in a separate issue:

1. The existing endpoints to show role mappings don't sort. When listing user role assignments hiding inherited roles, realm roles come first and then client roles, both are not sorted. When listing user roles including inherited roles, client roles come first and are sorted, then realm roles follow and are not sorted.
2. The existing endpoints to show role mappings don't respect fine-grained permissions. Hiding inherited roles shows all directly assigned roles (also without map-role permissions), showing inherited roles only shows client roles with map-role permission but still shows all realm roles.
3. The code around fine-grained permissions could really use a refactoring, there is a lot of code duplication in there.

All of this already exists in main and is not introduced by the changes here. The question is also, which behaviour is intended i.e. needing map-role permission to see a role as available definitely makes sense but I am not sure it should also be necessary to see a role that is assigned.

@vramik Can you please have a look?

ahus1 commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/AvailableRoleMappingResource.java:
```diff
@@ -176,8 +212,35 @@ public final List<ClientRole> listCompositeUserRoleMappings(@PathParam("id") Str
+    }
+
+    private List<ClientRole> searchForClientRolesByIds(RealmModel realm, Stream<String> includedIDs, String search, long first, long max) {
+        //TODO: not really sure about that casting of longs to ints here but I did not want to change the API sinature
```
Feel free to change  `listAvailableRoleMappings()` to have Integers to be consistent with `UserResource` (see below) to avoid the casting. This would resolve this and the following TODO.

```
 @QueryParam("first") Integer firstResult, @QueryParam("max") Integer maxResults,
```

ahus1 commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/AvailableRoleMappingResource.java:
```diff
@@ -176,8 +212,35 @@ public final List<ClientRole> listCompositeUserRoleMappings(@PathParam("id") Str
+    private Set<String> getRoleIdsWithPermissions(String roleResourceScope, String clientResourceScope) {
+        Set<String> roleIds = this.auth.roles().getRolesWithPermission(roleResourceScope);
+        Set<String> clientIds = this.auth.clients().getClientsWithPermission(clientResourceScope);
+        //TODO: this might be pretty inefficient, directly joining client IDs with role IDs in one query would be better
```
Please describe if you think this something you want to implement as part of this PR, or if it is good enough / as good as before, and you want handle this in a future issue/PR. 

Either way, there shouldn't be a TODO in the code for this. 

If you choose to have another issue for that, you can even reference the issue here to point to a discussion there.

sschu replied:
Thanks. I'll keep it as it is for now. The current code also populates the caches which has is upsides. I would revisit this should it turn out to be an issue in practice.

ahus1 reviewed (commented):
@sschu - thank you for this PR. I commented on some TODOs, I hope this helps until @vramik has a final look at the PR early next week if time permits.

In general I agree to keep PRs as small as possible, and move work to future PRs if the PR as is is an improvement.

sschu commented:
Thanks for the comments @ahus1 ! Waiting for @vramik (no pressure) ;)

vramik reviewed (changes requested):
Thank you @sschu! The PR looks great. I've several comments.

> * I added the new search method as a default method so alternative `RoleLookupProviders` still compile. However, they will encounter a runtime exception if they don't implement the method so this might not be ideal.

I agree this is not ideal and I believe we could leave it without default implementation. @ahus1 wdyt? If it turns out we have to have a default implementation, I'd advise to return either `null` or `Stream.empty()`, but from my POV the best would be to not have default implementation there.

> * Searching for available client roles in the `RoleLookupProvider` searches for matches in role name and client name because that's what the current API is doing to list available client roles. However, it is not consistent with what the other search functions are doing, they search in role name and role description.

True, @edewit is this desired behavior from UI point of view?

> * Listing effective role mappings doesn't offer pagination (it was like that before) but the admin console actually sends a request with pagination, at least when listing roles for a user.

Interesting, could you please link here a place in code from where the admin console sends the request? I'd say this is something we should look at, probably within follow up issue, could you please create one?

> * Listing effective role mappings could probably be more efficient by going to the DB directly similar to available role mappings. But it might not be that important considering that the amount of assigned roles should not be that big.

I agree, if you'd consider to make this more efficient, please do it within follow-up issue and PR.

> * The way this is currently going I see a lot of code duplication searching for realm/client/all roles. I am wondering if it would be better to just have an RoleQueryType enum with values CLIENT_ROLES,REALM_ROLES,ALL since inside the code differences are minimal.

You're right, we may want to refactor / improve the code in future, would you like to create `enhancement` issue for it?

ahus1 commented:
> from my POV the best would be to not have default implementation there.

I agree that a default implementation only makes sense if it returns a meaningful default, and returning an exception is no meaningful default IMHO. If there is no meaningful default implementation, don't provide one as @vramik suggested.

sschu commented on rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/RoleMappingResource.java:
```diff
@@ -2,42 +2,26 @@
 
 import static org.keycloak.admin.ui.rest.model.RoleMapper.convertToModel;
 
+import java.util.HashSet;
```
I deleted them.

sschu commented:
@vramik @ahus1 Thanks for the feedback! I included all remarks so from my perspective the PR is good to go if CI is green.

sschu commented:
@vramik With regards to the admin UI using pagination of listing effective role assignments: I assume this is the code but not sure, frontend isn´t my area:
https://github.com/keycloak/keycloak/blob/153d1a98418c4f9c6dce006cdf99f14fd6f48254/js/apps/admin-ui/src/components/role-mapping/resource.ts#L30-L42

edewit commented:
@sschu yep that is correct

sschu commented:
@vramik I don't think the test failure is created. I also created the follow-up issues you asked for.

vramik reviewed (approved):
Thank you @sschu!

ahus1 reviewed (approved):
Approving as of @vramik's review. Thank you to everyone involved in this PR to make this performance improvement happen!

ahus1 commented:
@edewit - could you please re-approve this PR for the UI team, as your last review is now stale after the lasted rework from the author. For the one GHA that failed due to flakiness I triggered a rerun.

Thanks!

sschu commented:
Cool, thanks a lot everybody!
